src/main/java/uk/gov/hmcts/ccd/domain/service/search/CaseSearchResultGenerator.java

```diff
@@ -0,0 +1,177 @@
+import uk.gov.hmcts.ccd.domain.service.aggregated.GetCaseTypeOperation;
+import uk.gov.hmcts.ccd.domain.service.aggregated.SearchQueryOperation;
+import uk.gov.hmcts.ccd.endpoint.exceptions.BadRequestException;
+import uk.gov.hmcts.ccd.endpoint.exceptions.BadSearchRequest;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static uk.gov.hmcts.ccd.domain.service.common.AccessControlService.CAN_READ;
+
+@Service
+public class CaseSearchResultGenerator {
+
+    private final UserRepository userRepository;
+    private final GetCaseTypeOperation getCaseTypeOperation;
+    private final SearchQueryOperation searchQueryOperation;
+
+    public CaseSearchResultGenerator(@Qualifier(CachedUserRepository.QUALIFIER) final UserRepository userRepository,
+                                     @Qualifier(AuthorisedGetCaseTypeOperation.QUALIFIER) final GetCaseTypeOperation getCaseTypeOperation,
+                                     final SearchQueryOperation searchQueryOperation) {
+        this.userRepository = userRepository;
+        this.getCaseTypeOperation = getCaseTypeOperation;
+        this.searchQueryOperation = searchQueryOperation;
+    }
+
+    public UICaseSearchResult execute(final String caseTypeId,
+                                      final CaseSearchResult caseSearchResult,
+                                      final String useCase) {
+        // TODO: Filter out fields from result that haven't been requested before returning (RDM-8556)
+        return new UICaseSearchResult(
+            buildHeaders(caseTypeId, useCase, caseSearchResult),
+            buildItems(useCase, caseSearchResult),
+            caseSearchResult.getTotal(),
+            useCase
+        );
+    }
+
+    private List<SearchResultViewItem> buildItems(String useCase, CaseSearchResult caseSearchResult) {
+        List<SearchResultViewItem> items = new ArrayList<>();
+        caseSearchResult.getCases().forEach(caseDetails -> {
+            getCaseTypeDefinition(caseDetails.getCaseTypeId()).ifPresent(caseType -> {
+                final SearchResult searchResult = searchQueryOperation.getSearchResultDefinition(caseType, useCase);
+                items.add(buildSearchResultViewItem(caseDetails, caseType, searchResult));
+            });
+        });
+
+        return items;
+    }
+
+    private List<SearchResultViewHeaderGroup> buildHeaders(String caseTypeId, String useCase, CaseSearchResult caseSearchResult) {
+        List<SearchResultViewHeaderGroup> headers = new ArrayList<>();
+        getCaseTypeDefinition(caseTypeId).ifPresent(caseType -> {
+            SearchResultViewHeaderGroup caseSearchHeader = buildHeader(useCase, caseSearchResult, caseTypeId, caseType);
+            headers.add(caseSearchHeader);
+        });
+
+        return headers;
+    }
+
+    private Optional<CaseTypeDefinition> getCaseTypeDefinition(String caseTypeId) {
+        return getCaseTypeOperation.execute(caseTypeId, CAN_READ);
```

Comment:
I wouldn't call the whole operation. I think we can simplify things by just getting it from the DefRepo? The external API will take care of access control. Important thing that we test this.